### PR TITLE
allow for postcopy migration

### PR DESCRIPTION
**Pull request by ashleyschuett**
Signed-off-by: Ashley Schuett <ashleyns1992@gmail.com>
**What this PR does / why we need it**:

This adds the option for users/operators to use posy copy migration strategy when migrating a VM. 

The operator can configure the system to use post migration as follows:
```
kind: KubeVirt 
spec: 
   configuration: 
       migrations:
            migrationMode: "PostCopy"
```

**Special notes for your reviewer**:

**Release note**:
```release-note
Allow for post copy VMI migrations
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/manager.go [hunk 1]
```diff
@@ -196,6 +196,7 @@ func (l *LibvirtDomainManager) initializeMigrationMetadata(vmi *v1.VirtualMachin
 	domainSpec.Metadata.KubeVirt.Migration = &api.MigrationMetadata{
 		UID:            vmi.Status.MigrationState.MigrationUID,
 		StartTimestamp: &now,
+		Mode:           v1.MigrationPreCopy,
```
Comment (xpivarc) [inline pkg/virt-launcher/virtwrap/manager.go:199]: We can set this from `MigrationState`. (Useful if we ever don't start with pre-copy)

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 2]
```diff
@@ -395,6 +401,17 @@ const (
 	MigrationAbortInProgress MigrationAbortStatus = "Aborting"
 )
 
+//
+// +k8s:openapi-gen=true
+type MigrationMode string
+
+const (
+	// MigrationPreCopy means the VMI migrations that is currenly running is in pre copy mode
+	MigrationPreCopy MigrationMode = "pre-copy"
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:410]: pre-copy/PreCopy and post-copy/PostCopy ? woudl that be more consistent with out other values work in our api?

File: tests/migration_test.go [hunk 3]
```diff
@@ -1073,18 +1073,191 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
+					"progressTimeout":         "10",
 					"completionTimeoutPerGiB": "5",
 					"bandwidthPerMigration":   "1Mi",
 				}
 				migrationData, err := json.Marshal(data)
 				Expect(err).ToNot(HaveOccurred())
 				tests.UpdateClusterConfigValueAndWait("migrations", string(migrationData))
 			})
-			PIt("[test_id:2227] should abort a vmi migration without progress", func() {
+			It("[test_id:2227] should abort a vmi migration without progress", func() {
```
Comment (xpivarc) [inline tests/migration_test.go:1260]: Did we fixed this flaky test?

File: pkg/virt-controller/watch/migration.go [hunk 4]
```diff
@@ -538,6 +538,7 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
 				TargetNode:   pod.Spec.NodeName,
 				SourceNode:   vmi.Status.NodeName,
 				TargetPod:    pod.Name,
+				Mode:         virtv1.MigrationPreCopy,
```
Comment (davidvossel) [inline pkg/virt-controller/watch/migration.go:541]: is it right that preCopy is set unconditionally here?

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 5]
```diff
@@ -1334,14 +1351,21 @@ type SMBiosConfiguration struct {
...
-	UnsafeMigrationOverride           bool               `json:"unsafeMigrationOverride,string"`
+	NodeDrainTaintKey                 *string `json:"nodeDrainTaintKey,omitempty"`
+	ParallelOutboundMigrationsPerNode *uint32 `json:"parallelOutboundMigrationsPerNode,string,omitempty"`
+	ParallelMigrationsPerCluster      *uint32 `json:"parallelMigrationsPerCluster,string,omitempty"`
+	*PerVMIMigrationConfiguration     `json:",inline"`
+}
+
+// PerVMIMigrationConfiguration holds migration options that can be configurated per vmi
+// +k8s:openapi-gen=true
+type PerVMIMigrationConfiguration struct {
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1362]: Is there a reason why we now want to expose all these other tunables per vmi?

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 6]
```diff
@@ -1334,14 +1351,21 @@ type SMBiosConfiguration struct {
...
+	AllowAutoConverge       *bool              `json:"allowAutoConverge,string,omitempty"`
+	BandwidthPerMigration   *resource.Quantity `json:"bandwidthPerMigration,omitempty"`
+	CompletionTimeoutPerGiB *int64             `json:"completionTimeoutPerGiB,string,omitempty"`
+	ProgressTimeout         *int64             `json:"progressTimeout,string,omitempty"`
+	UnsafeMigrationOverride *bool              `json:"unsafeMigrationOverride,string,omitempty"`
+	UsePostCopy             *bool              `json:"usePostCopy,string,omitempty"`
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1368]: should the new ```MigrationMode``` be used to indicate this now?

File: pkg/virt-launcher/virtwrap/postcopy.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+			err = l.updateVMIMigrationMode(dom, vmi, v1.MigrationPostCopy)
+			if err != nil {
+				log.Log.Object(vmi).Reason(err).Error("Unable to update migration mode on domain xml")
+			}
+		}
+	})
+}
+
+func vmiHasLocalStorage(vmi *v1.VirtualMachineInstance) bool {
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/postcopy.go:38]: I'm a bit confused as to why isn't this allowed?

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 8]
```diff
@@ -380,6 +384,8 @@ type VirtualMachineInstanceMigrationState struct {
 	AbortStatus MigrationAbortStatus `json:"abortStatus,omitempty"`
 	// The VirtualMachineInstanceMigration object associated with this migration
 	MigrationUID types.UID `json:"migrationUid,omitempty"`
+	// Lets us know if the vmi is currenly running pre or post copy migration
+	Mode MigrationMode `json:"mode,omitempty"`
```
[the remark this answers is not in the document]
Comment (vladikr) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:388]: @xpivarc  it's already there as `UsePostCopy` - this can only be either pre-copy or post-copy

File: pkg/virt-controller/watch/migration.go (hunk 4 above)
Comment (ashleyschuett) [reply pkg/virt-controller/watch/migration.go:541]: This `MigrationState` is creating the initial migration state and since all migrations start in precopy mode I thought it was most accurate to have the mode in this object reflect that. The mode could instead not be set during the initial creation of the migration state object, then once the migration starts have it set?

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 5 above)
Comment (ashleyschuett) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1362]: My thought was that if we assume that the user creating the VMI knows the most information about how the migration should be handled (i.e using post copy) it would also make sense for them to be able to configure the other migration options for their specific VM. Maybe that's giving them to much power though and they should only have the option between pre vs post copy. I don't have a strong opinion

File: pkg/virt-launcher/virtwrap/manager.go (hunk 1 above)
Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/manager.go:199]: We could, unless we decide it should be set only once the migration is started. Depends how we resolve https://github.com/kubevirt/kubevirt/pull/4016/files#r473326575 

If we decide to not set it there then this value would need to be explicitly set to pre/post/etc here

File: pkg/virt-launcher/virtwrap/postcopy.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/cli"
+
+	libvirt "libvirt.org/libvirt-go"
+)
+
+func (l *LibvirtDomainManager) registerIterationEventForPostCopy(vmi *v1.VirtualMachineInstance, dom cli.VirDomain, options *cmdclient.MigrationOptions, migrationError chan error) {
+	l.virConn.DomainEventMigrationIterationRegister(func(c *libvirt.Connect, d *libvirt.Domain, event *libvirt.DomainEventMigrationIteration) {
+		if vmiHasLocalStorage(vmi) {
```
[the remark this answers is not in the document]
Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/postcopy.go:15]: > Do we consider a pvc which can be backed by local storage?

I did not, I can add that to the check 

>Can this be evaluated in the creation time of VM/VMI?(This would give better feedback to user...)

Would the expected UX be the VM/VMI can not be created if this is the case? What if the system level migration strategy is set to postcopy? can vm/vmis with local storage not be created at all?

File: pkg/virt-launcher/virtwrap/postcopy.go (hunk 7 above)
Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/postcopy.go:38]: This was done for ease of a first pass implementation

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 6 above)
Comment (ashleyschuett) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1368]: do you mean instead of a boolean specify which migration to use? I thought it might make more sense to keep as a boolean since the flag to use postcopy in a migration is a boolean, but we could have the user explicitly set `PostCopy` or `PreCopy`. Should it then be required?

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:388]: Thanks @vladikr , so I would just ask to sync PR description with code.

File: pkg/virt-launcher/virtwrap/postcopy.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-launcher/virtwrap/postcopy.go:15]: I am not sure if there is a way to check if pvc local...

Yes, I thought that we could not allow it or fire an alert(if we can't be sure)... So there are two situations:
1. Cluster-wide setting is postcopy and VMI is not specified/set to postcopy 
2. Cluster-wide setting is precopy but VMI override this to postcopy

In both, I would be inclined to not allow it as the system can't satisfy it.

File: pkg/virt-launcher/virtwrap/postcopy.go (hunk 7 above)
Comment (vladikr) [reply pkg/virt-launcher/virtwrap/postcopy.go:38]: I see. Local disks are normally being block-migrated. We calculate whether the vmi should be block-migrated or not during the vmi creation [1] 
As far as I know, block-migration should work with post-copy. However, if these disks are missing from [1], we should just add it :)

[1] https://github.com/kubevirt/kubevirt/blob/a888cef3fac94366b6b5ee71e8496d2c19f9f51e/pkg/virt-handler/vm.go#L1798

File: pkg/virt-launcher/virtwrap/postcopy.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (vladikr) [reply pkg/virt-launcher/virtwrap/postcopy.go:15]: Currently, virt-handler decides whether VMI is `migratable` during the VMI start. When migration spec is posted, we can quickly reject it for non-migratable VMIs.
Please take a look at [1]

However, I think that `vmiHasLocalStorage` may not be necessary as I mentioned in another comment - but maybe I'm missing something. Please correct me if I'm wrong here :)

[1] https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L728

File: pkg/virt-controller/watch/migration.go (hunk 4 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/migration.go:541]: yeah, I think it would be more accurate to not set mode until we know what it is. For instance, If the UI ever displays the mode, and someone see's a value that don't expect there, it might be kind of confusing.

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 5 above)
Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1362]: @vladikr @rmohr what do you think here?

I remember when we were first talking about migrations, we considered all these tunables to be things only the administrator would have the ability to modify. and that the role of the user would have all this abstracted away.

If that's still our thinking, then the post/pre copy setting itself might not even belong on the the vmi api... however, this could be useful.

Maybe this isn't a discussion we should have during this PR and the VMI api should stay as it is today with no per VMI migration tunings. We can introduce postcopy at the cluster config level still and have separate discussion focused on what tunings belong on the VMI itself?

Thoughts?

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 6 above)
Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1368]: it could be left optional and we default to precopy when no value is set.

I only brought this up because of the "mode" section on the VMI status. It seems like we'd want to use consistent ways across the api for representing pre/post copy

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 5 above)
Comment (vladikr) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1362]: I agree. I also don't think we should expose most of these tunables to the user.
Perhaps it's fine to expose `AllowAutoConverge` and `UsePostCopy` - but I'm also fine with keeping it in the hands of the admin.

Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1362]: IMO one interesting question would be to ask: How can admin change migration options per vmi basis? (If that makes sense)

Comment (vladikr) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1362]: Most of the tunables in this section make sense only as a cluster wide setting. The only two I would consider being set per migration are auto-converge and post-copy.
Perhaps we should split this tunable into an admin setting (allow/forbid auto-converge/post-copy) and a user per migration setting (forceAutoConverge and usePostCopy)

What do you think?

Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1362]: yes, that might make sense.

For the sake of getting this feature merged, I suggest we keep everything at the admin level for now, and discuss the VMI API changes as a follow up. I don't think anyone questions the usefulness of post copy, which should make getting the fundamental logic in pretty straight forward.  I do think there are likely "pre vmi" migration tunables that can be exposed, but that likely goes beyond the scope of this initial feature implementation.

File: tests/migration_test.go [hunk 10]
```diff
@@ -1073,18 +1073,191 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
 		})
 
+		Context("migration postcopy", func() {
+			confirmMigrationMode := func(vmi *v1.VirtualMachineInstance, expectedMode v1.MigrationMode) {
+				By("Retrieving the VMI post migration")
+				vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Verifying the VMI's migration state")
+				Expect(vmi.Status.MigrationState.Mode).To(Equal(v1.MigrationPostCopy))
```
Comment (kbidarkar) [inline tests/migration_test.go:1083]: I think we want to use `expectedMode` here, instead of `v1.MigrationPostCopy`

File: tests/migration_test.go [hunk 11]
```diff
@@ -1073,18 +1073,191 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
+			}
+
+			It("should migration using cluster level config for postcopy", func() {
```
Comment (kbidarkar) [inline tests/migration_test.go:1086]: I think we want to change this to `s/migration/migrate/`

File: tests/migration_test.go [hunk 12]
```diff
@@ -1073,18 +1073,191 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
+				confirmMigrationMode(vmi, v1.MigrationPostCopy)
+
+				// delete VMI
+				By("Deleting the VMI")
+				Expect(virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})).To(Succeed())
+
+				By("Waiting for VMI to disappear")
+				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 240)
+			})
+			It("should migration still migrate with local storage as precopy", func() {
```
Comment (kbidarkar) [inline tests/migration_test.go:1126]: I think we want to change this to `s/should migration still/should still migrate/`

File: tests/migration_test.go [hunk 13]
```diff
@@ -1073,18 +1073,191 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
+				vmi = runVMIAndExpectLaunch(vmi, 240)
+
+				// execute a migration, wait for finalized state
+				By("Starting the Migration")
+				migration := tests.NewRandomMigration(vmi.Name, vmi.Namespace)
+				migrationUID := runMigrationAndExpectCompletion(migration, 180)
+
+				// check VMI, confirm migration state
+				confirmVMIPostMigration(vmi, migrationUID)
+				confirmMigrationMode(vmi, v1.MigrationPreCopy)
```
Comment (kbidarkar) [inline tests/migration_test.go:1147]: I was expecting this testcase to fail because of below mentioned reasons. 
This feels like false-positive.

- We passing expected value as `v1.MigrationPreCopy` and then currently in the test, anyways it is being checked whether the `vmi.Status.MigrationState.Mode` is  `v1.MigrationPostCopy` in `confirmMigrationMode`. So, this testcase should have failed actually.

Not sure, if I am missing something here, better to double check this once.

File: pkg/virt-launcher/virtwrap/postcopy.go (hunk 7 above)
Comment (vladikr) [reply pkg/virt-launcher/virtwrap/postcopy.go:33]: Is this now being using only in tests? If so, perhaps we should move it to the test itself?

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/postcopy.go:33]: I'm wondering, can't we block migrate with the emptyDisk? 
I understand that we have an issue migrating local volumes. However, these local storage volumes will be created with ReadWriteOnce access, so we will prevent the migration of a VMI with such a volume ahead of time in virt-handler. Is my assumption wrong?

**Comment by ashleyschuett** [comment]
The current PR has a test to ensure that the vmi is switch from pre copy to post copy after one iteration. However I have talk with others about trying to run a test that proves that post copy is able to migrate a load that pre copy would fail on. In order to do this I have tried to use the stress[1] test which we use currently but change it to run a more intense load. I haven't been able to find a load the ensures that the pre copy migration fails though. 

After talking through this with @vladikr it was suggested to instead measure the time it takes to migrate a vmi with load and test that the post copy migration is completed in less time than a pre copy migration. Again, I have been unable to find a way to do this consistently and am afraid of introducing a flaky test. I tried to do some research on how to test these two different migrations and in one reading found that pre copy should, for it's total migration time, take less time than post copy [2] if I am understanding this correctly.  

Is a test like this necessary or is the current testing enough? If not, what should be tested to ensure confidence in the post copy migration?  

[1] https://linux.die.net/man/1/stress
[2] https://core.ac.uk/download/pdf/287021688.pdf

File: pkg/virt-launcher/virtwrap/postcopy.go (hunk 7 above)
Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/postcopy.go:33]: From my understanding you are correct and this is not needed. I had deleted it, but it looks like I messed up when I rebased. I'll remove the functions, along with the tests for it.

**Comment by vladikr** [comment]
> The current PR has a test to ensure that the vmi is switch from pre copy to post copy after one iteration. However I have talk with others about trying to run a test that proves that post copy is able to migrate a load that pre copy would fail on. In order to do this I have tried to use the stress[1] test which we use currently but change it to run a more intense load. I haven't been able to find a load the ensures that the pre copy migration fails though.
> 
> After talking through this with @vladikr it was suggested to instead measure the time it takes to migrate a vmi with load and test that the post copy migration is completed in less time than a pre copy migration. Again, I have been unable to find a way to do this consistently and am afraid of introducing a flaky test. I tried to do some research on how to test these two different migrations and in one reading found that pre copy should, for it's total migration time, take less time than post copy [2] if I am understanding this correctly.
> 
> Is a test like this necessary or is the current testing enough? If not, what should be tested to ensure confidence in the post copy migration?
> 
> [1] https://linux.die.net/man/1/stress
> [2] https://core.ac.uk/download/pdf/287021688.pdf

I don't this we should test the performance of a post-copy migration vs a per-copy migration. Our focus should be on KubeVirts functionality, whether we handle the migration correctly or not
We should run the same tests we run for pre-copy for post-copy, except for the timeouts of the migration. These should be ignored for the post-copy migration entirely in the migration monitor.

**Comment by ashleyschuett** [comment]
> We should run the same tests we run for pre-copy for post-copy, except for the timeouts of the migration. These should be ignored for the post-copy migration entirely in the migration monitor. 

Do you mean we should run each migration test for both precopy and post copy? So instead of having a singular test to make sure that when post copy is set that the migration is ran using post copy, I should duplicate the 15 or so tests and have each one run for both pre and post?

**Comment by vladikr** [comment]
> > We should run the same tests we run for pre-copy for post-copy, except for the timeouts of the migration. These should be ignored for the post-copy migration entirely in the migration monitor.
> 
> Do you mean we should run each migration test for both precopy and post copy? So instead of having a singular test to make sure that when post copy is set that the migration is ran using post copy, I should duplicate the 15 or so tests and have each one run for both pre and post?

No, it was a general statement :)
Let's see what makes sense. We already test different combinations for pre-copy migration, most of it should work for post-copy.

From my pov, we should test a block migration and live migration.
Your current test will set the migration to block, so maybe reuse one of the NFS tests with post-copy.

I'm also wondering if cancelation works as expected.

File: pkg/virt-launcher/virtwrap/postcopy.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+	libvirt "libvirt.org/libvirt-go"
+)
+
+func (l *LibvirtDomainManager) registerIterationEventForPostCopy(vmi *v1.VirtualMachineInstance, dom cli.VirDomain, options *cmdclient.MigrationOptions, migrationError chan error) {
+	l.virConn.DomainEventMigrationIterationRegister(func(c *libvirt.Connect, d *libvirt.Domain, event *libvirt.DomainEventMigrationIteration) {
+		if event.Iteration == 1 {
+			log.Log.Object(vmi).Info("Signaled start for post copy migration")
+
+			err := dom.MigrateStartPostCopy(uint32(0))
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/postcopy.go:18]: @ashleyschuett could you please remind me why was this needed? Also, do we have a test for this scenario?

Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/postcopy.go:18]: All migrations must be started in PreCopy mode, once the migration is started it then needs to be switched to PostCopy.

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/postcopy.go:18]: yes... I'm just not so much familial with `.Iteration`. Are we waiting for migration start?
Why is this has to equal to `1`? What happens if we will miss one iteration?

I'm in general worried about libvirt events callbacks. if we are waiting for migration start, we could simply switch to post-copy in the monitoring goroutine.[1]
`libvirt.DOMAIN_JOB_UNBOUNDED` tells us that migration is running and we can switch here.

The advantage is that we actively monitor the sate of the migration in this thread and hopefully will not miss migration state changes.
Also, this is the place where we make all the migration decisions.

What do you think?


[1] https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/manager.go#L641

Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/postcopy.go:18]: I choose to watch for the iteration and switch after one iteration since that is how `virsh` implements it when `--postcopy-after-precopy` flag is specified for migration. But it does make sense to have this be more robust in case of an event not being sent from libvirt. 

If we move this to be handled in `liveMigrationMonitor` should we switch to postcopy as soon as an first `libvirt.DOMAIN_JOB_UNBOUNDED` is detected? Or should we still register callback for `DomainEventMigrationIterationRegister` in the  `liveMigraitonMonitor` and then switch on the first iteration or greater if the migration Mode is Pre when it should be Post?

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/postcopy.go:18]: > If we move this to be handled in `liveMigrationMonitor` should we switch to postcopy as soon as an first `libvirt.DOMAIN_JOB_UNBOUNDED` is detected?

iirc, `iteration` is iteration over RAM, after which all unlocked pages will be transferred. Without registering a callback, in our case, it would mean that we have observed progress (first time when we update `lastProgressUpdate`)

> Or should we still register callback for `DomainEventMigrationIterationRegister` in the `liveMigraitonMonitor` and then switch on the first iteration or greater if the migration Mode is Pre when it should be Post?

I don't think we should do both.

Moreover, I'm not confident that we want an immediate post-copy migration switch (unless there is a requirement to do so, that I'm not aware of)
Considering all the potential issues that may be associate with our implementation and post-copy migration  - I think the safest would be to switch to post-copy if we didn't manage to migrate in a reasonable amount of time, basically when `acceptableCompletionTime` is reached.
OpenStack has a similar logic [1]

[1] https://github.com/openstack/nova/blob/master/nova/virt/libvirt/driver.py#L9344

Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/postcopy.go:18]: I am updating this to work as described above. However, my concern is now with the way the mode is specified and will be shown to the user in the VMI status spec object and the UI. 

Right now the configuration key is called `MigrationMode` which can be specified to `PostCopy`. However the migration could complete in `PreCopy` mode if the `acceptableCompletionTime` is never reached and therefore the migration status object of the VMI will then show that the migration is completed and that the mode was `PreCopy` which could be confusing for users. 

I think because of this we should change the name of the configuration key. Maybe `PostCopyEnabled` or if we don't want to be so specific `MigrationModesAllowed` might be better names?  What do you think? Or do you think the naming already makes sense and users will understand?

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/postcopy.go]: That's a good point. I might be biased at this point, but I don't find this very confusing. But you're right a boolean may be clearer.
I like the `PostCopyEnabled`, but it may be confusing too, as it's not clear whether it is enabled right now or is it enabled in general. How about `PostCopyActive`? (I'm not against PostCopyEnabled, I'll be +1 if you'll choose this)

I went over the code again and I only have one doubt about the configuration field `spec.configuration.migrations.migrationMode: "PostCopy" itself.
Maybe instead of `migrationMode: "PostCopy" ` we should also switch to a boolean, `allowPostCopy` similar to `AllowAutoConverge`, as we don't force it now.
What do you think?
`

**Comment by gabrielecerami** [comment]
@ashleyschuett: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: pkg/virt-launcher/virtwrap/postcopy.go (hunk 14 above)
Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/postcopy.go]: > I went over the code again and I only have one doubt about the configuration field spec.configuration.migrations.migrationMode: PostCopy itself.
Maybe instead of migrationMode: PostCopy we should also switch to a boolean, allowPostCopy similar to AllowAutoConverge, as we don't force it now.
What do you think? 

Yes this is what i was trying to refer to, but did a poor job explaining. I think `allowPostCopy` is much more clear and will give a better UX experience.

File: pkg/virt-launcher/virtwrap/manager.go [hunk 15]
```diff
@@ -636,7 +643,9 @@ monitorLoop:
 			logger.Reason(err).Error("failed to get domain job info")
 			break
 		}
+
 		remainingData := int64(stats.DataRemaining)
+		fmt.Println(remainingData)
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/manager.go:648]: guess this is a debug leftover...

File: pkg/virt-launcher/virtwrap/manager.go [hunk 16]
```diff
@@ -648,6 +657,7 @@ monitorLoop:
 				progressWatermark = remainingData
 				lastProgressUpdate = now
 			}
+
 			// check if the migration is progressing
 			progressDelay := now - lastProgressUpdate
 			if progressTimeout != 0 &&
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/manager.go:663]: @ashleyschuett we should probably ignore this when we've switched to post-copy... what do you think?
I guess there is more harm than benefit.

File: tests/migration_test.go [hunk 17]
```diff
@@ -1011,6 +1020,78 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 				By("Waiting for NFS pod to disappear")
 				tests.WaitForPodToDisappearWithTimeout(tests.NFSTargetName, 120)
 			})
+
+			It(" should be migrated successfully, using guest agent on VM w/ postcopy", func() {
+				data := map[string]string{
+					"migrationMode":           "PostCopy",
```
Comment (vladikr) [inline tests/migration_test.go:1026]: ah, I guess this needs to be adjusted too.
Otherwise, the migration is being canceled.:

```
{"component":"virt-launcher","level":"error","msg":"operation aborted: migration out job: canceled by client","pos":"qemuMigrationJobCheckStatus:1563","subcomponent":"libvirt","thread":"34","timestamp":"2020-09-24T09:30:03.202000Z"}
```

File: pkg/virt-launcher/virtwrap/manager.go (hunk 16 above)
Comment (ashleyschuett) [reply pkg/virt-launcher/virtwrap/manager.go:662]: I was curious about this as well, once switched to post copy is there a chance the migration could for a time stop progressing but still be okay and should be continued? My concern with ignoring it is then all migration checks are ignored once the migration is switched to post and there's nothing monitoring it to cancel it. However, looking through the implementation from openstack that is also the case, so I gather that doesn't matter?

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/manager.go:662]: On one hand, you are right, we don't want to ignore all the checks, but the trouble with canceling the migration automatically in the post-copy is that we will lose state and I'm not sure it's helpful.
Perhaps we can log or emit an event instead of canceling?

Comment (davidvossel) [reply pkg/virt-launcher/virtwrap/manager.go:668]: id think that not timing out on progress post copy is probably fine. if the connection is completely severed, my expectation is that libvirt would fail the migration

**Review by davidvossel — COMMENTED** [review]
/approve

I'm sure it's possible we'll consider some further tunings at some point, but this looks like a reasonable start for post copy migration to me.

**Comment by vladikr** [comment]
Looks great! Thanks @ashleyschuett 
/lgtm
